### PR TITLE
Feature/safety fence

### PR DESCRIPTION
**Pull request by bhuvanchandra**
Working version of Safety Fence monitor and control.

### CODE REVIEW

File: safety_fence/msg/FenceDoorStatusArray.msg [hunk 1]
```diff
@@ -0,0 +1,4 @@
+std_msgs/Header header
+
+FenceDoorStatus[] statuses
+bool estop
```
Comment (Dale-Koenig) [inline safety_fence/msg/FenceDoorStatusArray.msg]: clarify this please

File: safety_fence/src/safety_fence_node.cpp [hunk 2]
```diff
@@ -0,0 +1,76 @@
+#include <stdio.h>
+#include <string>
+
+#include <ros/ros.h>
+
+#include <safety_fence/safety_fence.hpp>
+#include <rr_hw_interface/gpio/revpi_gpio.hpp>
+
+#include <safety_fence/FenceDoorStatusArray.h>
+#include <safety_fence/Indicator_LockControl.h>
```
Comment (Dale-Koenig) [inline safety_fence/src/safety_fence_node.cpp]: I don't get the naming convention here

File: safety_fence/src/safety_fence_node.cpp [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+        // std::cout << pin_name << " " << topic_name << std::endl;
+        rapyuta::RevPiGpioBoardConfig config;
+        pins.back().init(config);
+    }
+
+    ros::Rate loop_rate(scanFreq);
+    safety_fence::FenceDoorStatusArray status_msg;
+    status_msg.statuses.resize(doorNum);
+    ros::Publisher status_pub = n.advertise<safety_fence::FenceDoorStatusArray>(array_topic_name, 10);
+    ros::Subscriber indicator_lock_sub = n.subscribe("indicator_door_lock", 1000, indicatorDoorLockCallback);
```
Comment (Dale-Koenig) [inline safety_fence/src/safety_fence_node.cpp]: could probably make this 10 without problems

File: safety_fence/src/safety_fence_node.cpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    n.getParam("array_topic_name", array_topic_name);
+    n.getParam("door_num", doorNum);
+    n.getParam("pin_num", pinNum);
+    n.getParam("scan_freq", scanFreq);
+
+    for (int i = 0; i < pinNum; i++) {
+        std::string pin_num = "pin" + std::to_string(i);
+        n.getParam(pin_num + "/name", pin_name);
+        n.getParam(pin_num + "/topic_name", topic_name);
+        pins.push_back(rapyuta::safetyFence<rapyuta::RevPiGpio, rapyuta::RevPiGpioBoardConfig>(pin_name));
```
Comment (Dale-Koenig) [inline safety_fence/src/safety_fence_node.cpp]: pins.emplace_back (pin_name);

File: safety_fence/msg/Indicator_LockControl.msg [hunk 5]
```diff
@@ -0,0 +1,4 @@
+bool[] indicator_red     # True: ON, False: OFF
+bool[] indicator_green   # True: ON, False: OFF
+bool[] door_lock         # True: Deactivates lock, False: Activates lock
+
```
Comment (Dale-Koenig) [inline safety_fence/msg/Indicator_LockControl.msg]: Are we assuming all these are length two?

Comment (bhuvanchandra) [reply safety_fence/msg/Indicator_LockControl.msg]: yes

File: safety_fence/src/safety_fence_node.cpp (hunk 4 above)
Comment (bhuvanchandra) [reply safety_fence/src/safety_fence_node.cpp]: Ok

File: safety_fence/src/safety_fence_node.cpp (hunk 2 above)
Comment (bhuvanchandra) [reply safety_fence/src/safety_fence_node.cpp]: Renamed to FenceControl.h

File: safety_fence/src/safety_fence_node.cpp (hunk 3 above)
Comment (bhuvanchandra) [reply safety_fence/src/safety_fence_node.cpp]: ok

File: safety_fence/msg/FenceDoorStatusArray.msg (hunk 1 above)
Comment (bhuvanchandra) [reply safety_fence/msg/FenceDoorStatusArray.msg]: will update with appropriate comments.

File: safety_fence/src/safety_fence_node.cpp [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    n.getParam("array_topic_name", array_topic_name);
+    n.getParam("door_num", doorNum);
+    n.getParam("pin_num", pinNum);
+    n.getParam("scan_freq", scanFreq);
+
+    for (int i = 0; i < pinNum; i++) {
+        std::string pin_num = "pin" + std::to_string(i);
+        n.getParam(pin_num + "/name", pin_name);
+        n.getParam(pin_num + "/topic_name", topic_name);
+        pins.emplace_back(rapyuta::safetyFence<rapyuta::RevPiGpio, rapyuta::RevPiGpioBoardConfig>(pin_name));
```
Comment (Dale-Koenig) [inline safety_fence/src/safety_fence_node.cpp]:
```suggestion
        pins.emplace_back(pin_name);
```

File: safety_fence/launch/fence.launch [hunk 7]
```diff
@@ -0,0 +1,10 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<launch>
+  <arg name="host" default="RevPi6409" />
+  <arg name="usr" default="pi" />
+  <arg name="env-loader" default="/home/$(arg usr)/remote_env_loader.sh" />
+  <machine name="safety_fence_host" address="$(arg host)" user="$(arg usr)" env-loader="$(arg env-loader)" />
+  <node pkg="safety_fence" type="safety_fence" name="safety_fence" output="screen">
```
Comment (Dale-Koenig) [inline safety_fence/launch/fence.launch]:
```suggestion
  <node pkg="safety_fence" type="safety_fence" name="safety_fence" output="screen" machine="safety_fence_host">
```

File: safety_fence/launch/fence.launch [hunk 8]
```diff
@@ -0,0 +1,10 @@
...
+                <rosparam file="$(find safety_fence)/config/safety_fence_params.yaml" />
```
Comment (Dale-Koenig) [inline safety_fence/launch/fence.launch]:
```suggestion
    <rosparam file="$(find safety_fence)/config/safety_fence_params.yaml" />
```

**Pull request merged**